lib/src/puppetdb-cli.cc

```diff
@@ -107,7 +112,83 @@ PuppetDBConn::parseServerUrls(const json::JsonContainer& config) {
     }
 }
 
-size_t write_data(void *ptr, size_t size, size_t nmemb, FILE *stream) {
+class SynchronizedCharQueue {
+  public:
+    vector<char> front() {
+        boost::unique_lock<boost::mutex> mlock(mutex_);
+        cond_.wait(mlock, [&]{ return !queue_.empty(); });
+        return queue_.front();
+    }
+
+    vector<char> pop() {
+        boost::unique_lock<boost::mutex> mlock(mutex_);
+        cond_.wait(mlock, [&]{ return !queue_.empty(); });
+        auto item = queue_.front();
```

Comment:
I think this should be explicitly typed; `auto` will expand to `const vectorshit&` here, which *I think* can be invalidated by the pop() below. Better to make it a value explicitly, doing the copy on this line and trusting RVO to convert the return into a move.